syntax = "proto3";

package mds_rpc;

service MetadataService {
  rpc mkdir (MkdirRequest) returns (MkdirReply) {};
  rpc opendir (OpendirRequest) returns (OpendirReply) {};
  rpc rmdir(RmdirRequest) returns (RmdirReply) {};
  rpc stat (StatRequest) returns (StatReply) {};
  rpc readdir (ReaddirRequest) returns (ReaddirReply) {};

  rpc update_oss_cluster (UpdateOssClusterRequest) returns (UpdateOssClusterReply) {};
}



message FileStat {
  uint64 ino = 1;
  uint32 mode = 2;
  uint32 uid = 3;
  uint32 nlink = 4;
}

message DirRead {
  uint32 pos = 1;
  bytes pathname = 2;
}

message DirEntry {
  uint64 ino = 1;
  uint32 type = 2;
  bytes name = 3;
}


message MkdirRequest {
  uint64 uid = 1;
  uint64 rid = 2;
  uint32 mode = 3;
  bytes pathname = 4;
}

message MkdirReply {
  int32 ret_val = 1;
}

message OpendirRequest {
  uint64 uid = 1;
  uint64 rid = 2;
  bytes pathname = 3;
}

message OpendirReply {
  int32 ret_val = 1;
}

message RmdirRequest {
  uint64 uid = 1;
  uint64 rid = 2;
  bytes pathname = 3;
}

message RmdirReply {
  int32 ret_val = 1;
}

message StatRequest {
  uint64 uid = 1;
  uint64 rid = 2;
  bytes pathname = 3;
}

message StatReply {
  int32 ret_val = 1;
  FileStat stat = 2;
}

message ReaddirRequest {
  uint64 uid = 1;
  uint64 rid = 2;
  DirRead dir = 3;
}

message ReaddirReply {
  int32 ret_val = 1;
  DirEntry entry = 2;
}

message OssInfo {
  string name = 1;
  string addr = 2;
}

message UpdateOssClusterRequest {
  uint64 version = 1;
  repeated OssInfo info = 2;
}

message UpdateOssClusterReply {
  int32 ret_val = 1;
}